File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        
        stack = list()
        p = root
        
        while(len(stack) > 0 or p):
            if p:
                stack.append(p)
                p = p.left
            else:
                p = stack.pop()
                k = k - 1
                if k == 0:
                    return p.val
                p = p.right
                
        return -1
        
        

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type ListNode struct {
	Val  *TreeNode
	Next *ListNode
}

type Stack struct {
	Head *ListNode
}

func (s *Stack) Pop() *TreeNode {
	cur := s.Head
	s.Head = s.Head.Next
	cur.Next = nil
	return cur.Val
}
func (s *Stack) Push(n *TreeNode) {
	cur := &ListNode{
		Val:  n,
		Next: s.Head,
	}

	s.Head = cur
}
func (s *Stack) Empty() bool {
	return s.Head == nil
}

func kthSmallest(root *TreeNode, k int) int {
	stack := &Stack{}
	cur := root
	for cur != nil || !stack.Empty() {
		for cur != nil {
			stack.Push(cur)
			cur = cur.Left
		}
		cur = stack.Pop()
		k -= 1
		if k == 0 {
			break
		}
		cur = cur.Right
	}
	return cur.Val
}
